File: M8/Exemplos/App_eXperience_FireMonkey/Printer.pas
unit Printer;

interface
  uses
   Elgin.JNI.E1,Androidapi.Helpers,Androidapi.JNI.JavaTypes,Androidapi.JNI.GraphicsContentViewText,
   FMX.Graphics,FMX.Surfaces,FMX.Helpers.Android,System.SysUtils,System.IOUtils;


  const
    QTD_LINHAS_I9 = 10;
    QTD_LINHAS_INTERNA = 40;


  type

      BARCODE_TYPE = (UPC_A,UPC_E,EAN_13,JAN_13, EAN_8, JAN_8,CODE_39,ITF, CODE_BAR, CODE_93, CODE_128,QR_CODE);

      TPrinter = class
      private

      context : JContext;
      qtd_linhas : integer;

      function codeOfBarCode(barCodeName: BARCODE_TYPE): integer;

      public

      constructor Create(activity: JContext);
      function PrinterExternalImpStart(ip: string; port: integer): integer;
      function printerInternalImpStart(): integer;
      procedure PrinterStop;
      function  StatusSensorPapel(): integer;
      function  StatusGaveta() : integer;
      function  AvancaLinhas(quantLinhas : integer) : integer;
      function  CutPaper(quantLinhas : integer) : integer;
      function  ImprimeTexto(text, align, font: string; fontSize: integer;
             isCutPaper, isUnderline, isBold: boolean): integer;
      function ImprimeBarCode(barcodetype: BARCODE_TYPE; text, align: string; height,
  width: integer; isCutPaper: boolean): integer;
      function ImprimeQR_CODE(text, align : string; qrSize: integer; isCutPaper: boolean): integer;
      function ImprimeImagem(bitmap : TBitmap; isCutPaper: boolean) : integer;
      function ImprimeXMLNFCe(xmlNFCe, csc : string; indexcsc, param : integer; isCutPaper : boolean ): integer;
      function ImprimeXMLSAT(xmlSAT : string; param : integer; isCutPaper : boolean): integer;

    end;
  var
  Impressora : TPrinter;
implementation

{ TPrinter }

function TPrinter.AvancaLinhas(quantLinhas: integer): integer;
begin
   Result := TJTermica.JavaClass.AvancaPapel(quantLinhas);
end;

function TPrinter.codeOfBarCode(barCodeName: BARCODE_TYPE): integer;
begin
  case barCodeName of
    UPC_A: Result := 0;
    UPC_E: Result := 1;
    EAN_13: Result := 2;
    JAN_13: Result := 2;
    EAN_8: Result := 3;
    JAN_8: Result := 3;
    CODE_39: Result := 4;
    ITF: Result := 5;
    CODE_BAR: Result := 6;
    CODE_93: Result := 7;
    CODE_128: Result := 8;
    else
        Result := 0;
  end;
end;

constructor TPrinter.Create(activity: JContext);
begin
   context := activity;
   TJTermica.JavaClass.setContext(context);
end;

function TPrinter.CutPaper(quantLinhas: integer): integer;
begin
  Result := TJTermica.JavaClass.Corte(quantLinhas);
end;

function TPrinter.imprimeBarCode(barcodetype: BARCODE_TYPE; text, align: string; height,
  width: integer; isCutPaper: boolean): integer;
var
hri,alignValue : integer;
begin
 hri := 4; //NO PRINT

 alignValue := 0;

   if align = 'Esquerda' then
      alignValue := 0
   else if align = 'Centralizado' then
      alignValue := 1
   else
     alignValue := 2;

   TJTermica.JavaClass.DefinePosicao(alignValue);

   Result := TJTermica.JavaClass.ImpressaoCodigoBarras(codeOfBarCode(barCodeType),StringToJString(text), height, width, hri);

   if isCutPaper then
         CutPaper(qtd_linhas);

end;

// converte delphi bitmap para java bitmap
function BitmapToJBitmap(const ABitmap: TBitmap): JBitmap;
var
  LSurface: TBitmapSurface;
begin
  Result := TJBitmap.JavaClass.createBitmap(ABitmap.Width, ABitmap.Height, TJBitmap_Config.JavaClass.ARGB_8888);
  LSurface := TBitmapSurface.Create;
  try
    LSurface.Assign(ABitmap);
    SurfaceToJBitmap(LSurface, Result);
  finally
    LSurface.Free;
  end;
end;

function TPrinter.imprimeImagem(bitmap: TBitmap; isCutPaper: boolean): integer;
begin

  Result := TJTermica.JavaClass.ImprimeBitmap(BitmapToJBitmap(bitmap));

  //Result := TJTermica.JavaClass.ImprimeImagemMemoria(stringtojstring(System.IOUtils.TPath.GetDocumentsPath + PathDelim  + 'barcode_icon.png'),0);

  if isCutPaper then
         CutPaper(qtd_linhas);
end;

function TPrinter.imprimeQR_CODE(text, align: string; qrSize: integer;
  isCutPaper: boolean): integer;
var
alignValue : integer;
begin

    alignValue := 0;

   if align = 'Esquerda' then
      alignValue := 0
   else if align = 'Centralizado' then
      alignValue := 1
   else
     alignValue := 2;

   TJTermica.JavaClass.DefinePosicao(alignValue);

  Result := TJTermica.JavaClass.ImpressaoQRCode(StringToJString(text), qrSize, 2);

  if isCutPaper then
         CutPaper(qtd_linhas);
end;

function TPrinter.ImprimeTexto(text, align, font: string; fontSize: integer;
  isCutPaper, isUnderline, isBold: boolean): integer;
var
 alignValue,styleValue : integer;
begin
   alignValue := 0;
   styleValue := 0;

   if align = 'Esquerda' then
      alignValue := 0
   else if align = 'Centralizado' then
      alignValue := 1
   else
     alignValue := 2;

   if font = 'FONT B' then
      styleValue := 1;

   if isUnderline then
      styleValue := styleValue + 2;

   if isBold then
      styleValue := styleValue + 8;

   Result := TJTermica.JavaClass.ImpressaoTexto(StringToJString(text), alignValue, styleValue, fontSize);

   TJTermica.JavaClass.AvancaPapel(1);

   if isCutPaper then
         CutPaper(qtd_linhas);

end;

function TPrinter.imprimeXMLNFCe(xmlNFCe, csc: string; indexcsc,
  param: integer; isCutPaper : boolean): integer;
begin
   TJTermica.JavaClass.ImprimeXMLNFCe(StringToJString(xmlNFCe),indexcsc,StringToJString(csc), param);
   if isCutPaper then
         CutPaper(qtd_linhas);
end;

function TPrinter.imprimeXMLSAT(xmlSAT: string; param: integer; isCutPaper : boolean): integer;
begin
    TJTermica.JavaClass.ImprimeXMLSAT(StringToJString(xmlSAT), param);
    if isCutPaper then
         CutPaper(qtd_linhas);
end;

function TPrinter.printerExternalImpStart(ip: string; port: integer): integer;
begin
    printerStop;
    qtd_linhas := QTD_LINHAS_I9;
    try
        Result := TJTermica.JavaClass.AbreConexaoImpressora(3,StringToJString('i9'),StringToJString(ip), port);
    except
      on e: exception do
         printerInternalImpStart;
    end;

end;

function TPrinter.printerInternalImpStart;
begin
    printerStop;
    qtd_linhas := QTD_LINHAS_INTERNA;
    Result := TJTermica.JavaClass.AbreConexaoImpressora(6, StringToJString('M8'),StringToJString('USB'), 0);

end;

procedure TPrinter.printerStop;
begin
   TJTermica.JavaClass.FechaConexaoImpressora();
end;

function TPrinter.StatusGaveta: integer;
begin
  Result := TJTermica.JavaClass.StatusImpressora(1);
end;

function TPrinter.statusSensorPapel: integer;
begin
   Result := TJTermica.JavaClass.StatusImpressora(3);
end;



initialization

 Impressora := TPrinter.Create(TAndroidHelper.Context);

end.

File: Exemplos/App_eXperience_XamarinForms/Elgin_AppExperience_XamarinForms-font/M8XamarinForms.Android/Printer.cs
using Android.App;
using Android.Content;
using Android.Graphics;
using Android.Util;
using Com.Elgin.E1.Impressora;
using Java.IO;
using System;
using System.Collections.Generic;
using System.IO;


namespace M8XamarinForms
{
    public class Printer
    {
        private Context mContext;

        //Variavel utilizada para a verificação de se a conexão atual é por impressora interna ou não
        public bool isPrinterInternSelected;

        public Printer()
        {
            //mActivity = activity;
            //Termica.SetContext(mActivity);
        }

        public void SetActivity(Context context)
        {
            mContext = context;

            Termica.SetContext(context);
        }
        public int PrinterInternalImpStart()
        {
            PrinterStop();
            int result = Termica.AbreConexaoImpressora(6, "M8", "", 0);

            if (result == 0) isPrinterInternSelected = true;

            return result;
        }

        public int PrinterExternalImpStartByIP(string model, string ip, int port)
        {
            PrinterStop();
            try {
                int result = Termica.AbreConexaoImpressora(3, model, ip, port);

                if (result == 0) isPrinterInternSelected = false;

                return result;
            }
            catch (Exception)
            {
                return PrinterInternalImpStart();
            }
        }

        public int PrinterExternalImpStartByUSB(string model)
        {
            PrinterStop();
            try
            {
                int result = Termica.AbreConexaoImpressora(1, model, "USB", 0);

                if (result == 0) isPrinterInternSelected = false;

                return result;
            }
            catch (Exception)
            {
                return PrinterInternalImpStart();
            }
        }

        public void PrinterStop()
        {
            Termica.FechaConexaoImpressora();
        }

        public int AvancaLinhas(int linesNumber)
        {
            Log.Debug("PRINTER", "Antes de avançar " + linesNumber + " linhas");
            return Termica.AvancaPapel(linesNumber);
            Log.Debug("PRINTER", "Depois de avançar " + linesNumber + " linhas");
        }

        public void JumpLine()
        {
            int quant;
            //Se a impressão for por impressora externa, 5 é o suficiente; 15 caso contrário
            Log.Debug("PRINTER", "É impressora interna: " + isPrinterInternSelected);
            if (!isPrinterInternSelected) { quant = 5; }
            else quant = 15;

            AvancaLinhas(quant);
        }

        public int CutPaper(int cut)
        {
            return Termica.Corte(cut);
        }

        private int CodeOfBarCode(String barCodeName)
        {
            if (barCodeName.Equals("UPC-A"))
                return 0;
            else if (barCodeName.Equals("UPC-E"))
                return 1;
            else if (barCodeName.Equals("EAN 13") || barCodeName.Equals("JAN 13"))
                return 2;

            else if (barCodeName.Equals("EAN 8") || barCodeName.Equals("JAN 8"))
                return 3;
            else if (barCodeName.Equals("CODE 39"))
                return 4;
            else if (barCodeName.Equals("ITF"))
                return 5;
            else if (barCodeName.Equals("CODE BAR"))
                return 6;
            else if (barCodeName.Equals("CODE 93"))
                return 7;
            else if (barCodeName.Equals("CODE 128"))
                return 8;
            else return 0;
        }

        public int ImprimeBarCode(Dictionary<string, string> dictionary)
        {

            int barCodeType = CodeOfBarCode((String)dictionary["barCodeType"]);
            String text = (String) dictionary["text"];
            int height = (int) Int32.Parse(dictionary["height"]);
            int width = (int) Int32.Parse(dictionary["width"]);
            String align = (String) dictionary["align"];

            int hri = 4; // NO PRINT
            int result;
            int alignValue;

            if (align.Equals("Esquerda"))
            {
                alignValue = 0;
            }
            else if (align.Equals("Centralizado"))
            {
                alignValue = 1;
            }
            else
            {
                alignValue = 2;
            }

            Termica.DefinePosicao(alignValue);

            result = Termica.ImpressaoCodigoBarras(barCodeType, text, height, width, hri);
            return result;
        }

        public int ImprimeQR_CODE(Dictionary<string, string> dictionary)
        {
            int size = (int) Int32.Parse(dictionary["qrSize"]);
            String text = (String) dictionary["text"];
            String align = (String) dictionary["align"];

            int nivelCorrecao = 2;
            int result;
            int alignValue;

            if (align.Equals("Esquerda"))
            {
                alignValue = 0;
            }
            else if (align.Equals("Centralizado"))
            {
                alignValue = 1;
            }
            else
            {
                alignValue = 2;
            }

            Termica.DefinePosicao(alignValue);

            result = Termica.ImpressaoQRCode(text, size, nivelCorrecao);
            return result;
        }

        public int ImprimeXMLNFCe(Dictionary<string, object> map)
        {
            string xml = (string)map["xmlNFCe"];
            int indexcsc = (int)map["indexcsc"];
            string csc = (string)map["csc"];
            int param = (int)map["param"];
            return Termica.ImprimeXMLNFCe(xml, indexcsc, csc, param);
        }

        public int ImprimeXMLSAT(Dictionary<string, object> map)
        {
            string xml = (string)map["xmlSAT"];
            int param = (int)map["param"];
            return Termica.ImprimeXMLSAT(xml, param);
        }

        public int StatusGaveta()
        {
            return Termica.StatusImpressora(1);
        }

        public int AbrirGaveta()
        {
            return Termica.AbreGavetaElgin();
        }

        public int StatusSensorPapel()
        {
            return Termica.StatusImpressora(3);
        }

        public int ImprimeCupomTEF(Dictionary<string, string> dictionary)
        {
            string base64 = dictionary["base64"]; 

            return Termica.ImprimeCupomTEF(base64);
        }

        public int ImprimeTexto(Dictionary<string, string> dictionary)
        {
            string text = dictionary["text"];
            string align = dictionary["align"];
            string font = dictionary["font"];
            int fontSize = int.Parse(dictionary["fontSize"]);
            bool isBold = Convert.ToBoolean(dictionary["isBold"]);
            bool isUnderline = Convert.ToBoolean(dictionary["isUnderline"]);

            int result;
            int styleValue = 0;

            // ALINHAMENTO VALUE
            int alignValue;
            if (align.Equals("Esquerda"))
            {
                alignValue = 0;
            }
            else if (align.Equals("Centralizado"))
            {
                alignValue = 1;
            }
            else
            {
                alignValue = 2;
            }

            // ESTILO VALUE
            if (font.Equals("FONT B"))
            {
                styleValue += 1;
            }
            if (isUnderline)
            {
                styleValue += 2;
            }
            if (isBold)
            {
                styleValue += 8;
            }

            result = Termica.ImpressaoTexto(text, alignValue, styleValue, fontSize);
            return result;
        }

        public int ImprimeImagemPadrao()
        {
            Bitmap bitmap;
            int id = mContext.ApplicationContext.Resources.GetIdentifier("elgin_logo_default_print_image", "drawable", mContext.ApplicationContext.PackageName);
            bitmap = BitmapFactory.DecodeResource(mContext.ApplicationContext.Resources, id);

            int result;
            //Verifica se o método de impressão atual é por impressora interna ou externa e utiliza a função adequada para cada um
            if (isPrinterInternSelected) result = Termica.ImprimeBitmap(bitmap);
            else
            {
                result = Termica.ImprimeImagem(bitmap);
            }
            return result;
        }

        public int ImprimeImagem(Stream stream)
        {
            int result;

            Bitmap bitmap = BitmapFactory.DecodeStream(stream);

            //Verifica se o método de impressão atual é por impressora interna ou externa e utiliza a função adequada para cada um
            if (isPrinterInternSelected) result = Termica.ImprimeBitmap(bitmap);
            else
            {
                result = Termica.ImprimeImagem(bitmap);
            }
            return result;
        }

        public string CarregarArquivo(string nomeArquivo)
        {
            var stream = mContext.Resources.OpenRawResource(mContext.Resources.GetIdentifier(nomeArquivo, "raw", mContext.PackageName));
            string conteudoArquivo = "";
            using (var reader = new StreamReader(stream))
            {
                conteudoArquivo = reader.ReadToEnd();
            }
            return conteudoArquivo;
        }

    }

}